pkg/controller/cluster/cluster_controller_phases.go

```diff
@@ -0,0 +1,212 @@
+	"sigs.k8s.io/cluster-api/pkg/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+func (r *ReconcileCluster) reconcile(ctx context.Context, cluster *v1alpha2.Cluster) (err error) {
+	// TODO(vincepri): These can be generalized with an interface and possibly a for loop.
+	errors := []error{}
+	errors = append(errors, r.reconcileInfrastructure(ctx, cluster))
+	errors = append(errors, r.reconcilePhase(ctx, cluster))
+
+	// Determine the return error, giving precedence to the first non-nil and non-requeueAfter errors.
+	for _, e := range errors {
+		if e == nil {
+			continue
+		}
+		if err == nil || capierrors.IsRequeueAfter(err) {
+			err = e
+		}
+	}
+	return err
+}
+
+func (r *ReconcileCluster) reconcilePhase(ctx context.Context, cluster *v1alpha2.Cluster) error {
+	// Set the phase to "pending" if nil.
+	if cluster.Status.Phase == nil {
+		cluster.Status.SetTypedPhase(v1alpha2.ClusterPhasePending)
+	}
+
+	// Set the phase to "provisioning" if the Cluster has an InfrastructureRef object associated.
+	if cluster.Spec.InfrastructureRef != nil {
+		cluster.Status.SetTypedPhase(v1alpha2.ClusterPhaseProvisioning)
+	}
+
+	// Set the phase to "provisioned" if the infrastructure is ready.
+	if cluster.Status.InfrastructureReady {
+		cluster.Status.SetTypedPhase(v1alpha2.ClusterPhaseProvisioned)
+	}
+
+	// Set the phase to "failed" if any of Status.ErrorReason or Status.ErrorMessage is not-nil.
+	if cluster.Status.ErrorReason != nil || cluster.Status.ErrorMessage != nil {
+		cluster.Status.SetTypedPhase(v1alpha2.ClusterPhaseFailed)
+	}
+
+	// Set the phase to "deleting" if the deletion timestamp is set.
+	if !cluster.DeletionTimestamp.IsZero() {
+		cluster.Status.SetTypedPhase(v1alpha2.ClusterPhaseDeleting)
+	}
+
+	return nil
+}
+
+// reconcileExternal handles generic unstructured objects referenced by a Cluster.
+func (r *ReconcileCluster) reconcileExternal(ctx context.Context, cluster *v1alpha2.Cluster, ref *corev1.ObjectReference) (*unstructured.Unstructured, error) {
+	obj, err := external.Get(r.Client, ref, cluster.Namespace)
+	if err != nil {
+		if apierrors.IsNotFound(err) && !cluster.DeletionTimestamp.IsZero() {
+			return nil, nil
+		} else if apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second},
+				"could not find %v %q for Cluster %q in namespace %q, requeuing",
+				ref.GroupVersionKind(), ref.Name, cluster.Name, cluster.Namespace)
+		}
+		return nil, err
+	}
+
+	objPatch := client.MergeFrom(obj.DeepCopy())
+
+	// Delete the external object if the Cluster is being deleted.
+	if !cluster.DeletionTimestamp.IsZero() {
+		if err := r.Delete(ctx, obj); err != nil {
+			return nil, errors.Wrapf(err,
+				"failed to delete %v %q for Cluster %q in namespace %q",
+				obj.GroupVersionKind(), ref.Name, cluster.Name, cluster.Namespace)
+		}
+		return obj, nil
```

Comment:
Do we need to return obj here?

Reply:
Why not? The caller would need it for other things, for example the infrastructure reconciliation function uses it to get some fields